### PR TITLE
feat: add more precedence rules for route matching

### PR DESCRIPTION
**Pull request by Charlie17Li**
### Ⅰ. Describe what this PR did
add more precedence rules for route matching

### CODE REVIEW

File: pkg/ingress/kube/common/tool.go [hunk 1]
```diff
@@ -193,7 +196,29 @@ func SortHTTPRoutes(routes []*WrapperHTTPRoute) {
...
-			return len(routes[i].OriginPath) > len(routes[j].OriginPath)
+			if n, m := len(routes[i].OriginPath), len(routes[j].OriginPath); n != m {
+				return n > m
+			}
+
+			m1, m2 := routes[i].HTTPRoute.Match[0], routes[j].HTTPRoute.Match[0]
+			// methods
+			if n, m := len(m1.Method.GetRegex()), len(m2.Method.GetRegex()); true {
+				if n != 0 && m != 0 {
+					return n < m
```
Comment (johnlanni) [inline pkg/ingress/kube/common/tool.go:207]: When n==m, we need check other conditions.

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: pkg/ingress/kube/common/tool.go (hunk 1 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/common/tool.go:207]: ok

File: pkg/ingress/kube/common/tool.go [hunk 2]
```diff
@@ -193,7 +196,30 @@ func SortHTTPRoutes(routes []*WrapperHTTPRoute) {
...
+			m1, m2 := routes[i].HTTPRoute.Match[0], routes[j].HTTPRoute.Match[0]
+			// methods
+			if n, m := len(m1.Method.GetRegex()), len(m2.Method.GetRegex()); n != m {
+				if n != 0 && m != 0 {
+					return n < m
+				} else if n != 0 {
+					return true
+				} else if m != 0 {
+					return false
+				}
```
Comment (johnlanni) [inline pkg/ingress/kube/common/tool.go:212]: else { return n!=0 }

**Review by johnlanni — CHANGES_REQUESTED** [review]

**Comment by Xunzhuo** [comment]
@Charlie17Li Need to resolve some conflicts.

File: test/ingress/conformance/tests/httproute-same-host-and-path.yaml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  rules:
+    - http:
+        paths:
+          - pathType: Prefix
+            path: "/hello-world"
+            backend:
+              service:
+                name: infra-backend-v3
+                port:
+                  number: 8080
```
Comment (Xunzhuo) [inline test/ingress/conformance/tests/httproute-same-host-and-path.yaml:100]: nit: need a line at end.

File: pkg/ingress/kube/common/tool.go [hunk 4]
```diff
@@ -193,7 +196,27 @@ func SortHTTPRoutes(routes []*WrapperHTTPRoute) {
 		}
 
 		if routes[i].OriginPathType == routes[j].OriginPathType {
-			return len(routes[i].OriginPath) > len(routes[j].OriginPath)
+			if n, m := len(routes[i].OriginPath), len(routes[j].OriginPath); n != m {
```
Comment (Xunzhuo) [inline pkg/ingress/kube/common/tool.go:199]: I do not prefer the names `n` and `m`, they are lack of readability.

File: pkg/ingress/kube/common/tool.go [hunk 5]
```diff
@@ -193,7 +196,27 @@ func SortHTTPRoutes(routes []*WrapperHTTPRoute) {
...
+				return n > m
+			}
+
+			m1, m2 := routes[i].HTTPRoute.Match[0], routes[j].HTTPRoute.Match[0]
```
Comment (Xunzhuo) [inline pkg/ingress/kube/common/tool.go:203]: here too.

File: pkg/ingress/kube/ingress/controller.go [hunk 6]
```diff
@@ -568,11 +581,13 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
 
 			// host and path overlay check across different ingresses.
 			hostAndPath := wrapperHttpRoute.BasePathFormat()
-			if preIngress, exist := convertOptions.HostAndPath2Ingress[hostAndPath]; exist {
+			arr = append(arr, hostAndPath)
+			key := strings.Join(arr, ",")
+			if preIngress, exist := convertOptions.HostAndPath2Ingress[key]; exist {
```
Comment (johnlanni) [inline pkg/ingress/kube/ingress/controller.go:586]: This map is no longer “HostAndPath2Ingress”.

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: test/ingress/conformance/tests/httproute-same-host-and-path.yaml (hunk 3 above)
Comment (Charlie17Li) [reply test/ingress/conformance/tests/httproute-same-host-and-path.yaml:100]: ok

File: pkg/ingress/kube/common/tool.go (hunk 4 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/common/tool.go:199]: I see. how about `len1`, `len2` or `leni`, `lenj` or `pathLen1`, `pathLen2`?

File: pkg/ingress/kube/ingress/controller.go (hunk 6 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/ingress/controller.go:586]: ok, I've reacotor it.

**Review by Xunzhuo — APPROVED** [review]
Generally LGTM, thanks!

**Comment by johnlanni** [comment]
cc @SpecialYang

**Review by johnlanni — APPROVED** [review]
LGTM

**Pull request merged**